### PR TITLE
Fixes #20480: Fixed RegexNotFoundError on RTP channel 

### LINKED ISSUE #20480: RegexNotFoundError on RTP channel
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2019.03.18*. If it's not, read [this FAQ entry](https://github.com/ytdl-org/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [ X] I've **verified** and **I assure** that I'm running youtube-dl **2019.03.18**

### Before submitting an *issue* make sure you have:
- [x ] At least skimmed through the [README](https://github.com/ytdl-org/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/ytdl-org/youtube-dl#faq) and [BUGS](https://github.com/ytdl-org/youtube-dl#bugs) sections
- [ x] [Searched](https://github.com/ytdl-org/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x ] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [ x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```

```
paul@paul-NC10 ~ $ youtube-dl --verbose -o "/home/paul/Videos/%(title)s.%(ext)s" --fragment-retries 1 https://www.rtp.pt/play/p5285/e397170/jornal-da-tarde-2019 
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'-o', u'/home/paul/Videos/%(title)s.%(ext)s', u'--fragment-retries', u'1', u'https://www.rtp.pt/play/p5285/e397170/jornal-da-tarde-2019']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.03.18
[debug] Python version 2.7.12 (CPython) - Linux-4.4.0-143-generic-i686-with-LinuxMint-18.3-sylvia
[debug] exe versions: ffmpeg 3.2.6-1, ffprobe 3.2.6-1, phantomjs 2.1.1, rtmpdump 2.4-n111-gitfa8646d-ppa10p1
[debug] Proxy map: {}
[RTP] e397170: Downloading webpage
ERROR: Unable to extract player config; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 794, in extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 529, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/rtp.py", line 40, in _real_extract
    r'(?s)RTPPLAY\.player\.newPlayer\(\s*(\{.*?\})\s*\)', webpage, 'player config')
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 1004, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
RegexNotFoundError: Unable to extract player config; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://www.youtube.com/watch?v=BaW_jenozKc
- Single video: https://youtu.be/BaW_jenozKc
- Playlist: https://www.youtube.com/playlist?list=PL4lCao7KL_QFVb7Iudeipvc2BCavECqzc

Note that **youtube-dl does not support sites dedicated to [copyright infringement](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free)**. In order for site support request to be accepted all provided example URLs should not violate any copyrights.

---

### Description of your *issue*, suggested solution and other information

I'm using version 2019.03.18. I tried to download the programme I've pasted from The RTP website but got the error message above. Downloads have worked fine in the past. Whilst the programme will stream perfectly in a browser, I prefer to download. I'm afraid that's about all I can tell you.

### PR DESCRIPTION
**Pull request by rolyli**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x ] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

Site layout changed for RTP and many of the values for playlist descriptions are obsolete. Added a new regex search and proper json parsing.

### CODE REVIEW

**Comment by fiatjaf** [comment]
I tried this on a list of audios (https://www.rtp.pt/play/p510/e401638/aleixo-fm) and only got the first one. How can we make it download all them?

**Comment by nimeir** [comment]
So it seems like I can get a URL for the videos with a request like 
`https://www.rtp.pt/play/bg_l_ep/?listDate=&listQuery=&listProgram=510&listcategory=&listchannel=&type=radio&page=1` but it only shows 12 items per 'page'. Also, my browser makes a request for the page after the last page that returns no result. Perhaps the RTP website script knows when to stop making requests for further URLs when theres no more items returned?

 I could have the extractor go through all 20 or something pages but this seems makeshift. Does anyone have a good way of going about this?

**Comment by nimeir** [comment]
I finished adding a finctionality for playlists as well. If you specify the url as something like this https://www.rtp.pt/play/p510

Sorry the code isn't as clean as it could be as I'm new but I will fix try to fix anything if someone points out.

**Comment by fiatjaf** [comment]
I confirm this is working marvelously.

```
fiatjaf@menger ~/c/youtube-dl (rtp-fix)> python3 -m youtube_dl https://www.rtp.pt/play/p2611
[RTPPlaylist] 2611: Downloading webpage
[RTPPlaylist] 2611: Downloading webpage
[RTPPlaylist] 2611: Downloading webpage
[RTPPlaylist] 2611: Downloading webpage
[RTPPlaylist] 2611: Downloading webpage
[download] Downloading playlist: 2611
[RTPPlaylist] playlist 2611: Collected 32 video ids (downloading 32 of them)
[download] Downloading video 1 of 32
[RTP] e247140: Downloading webpage
[download] Destination: Aleixo Olímpico-e247140.mp3
[download] 100% of 3.54MiB in 00:02
[download] Downloading video 2 of 32
[RTP] e247145: Downloading webpage
[download] Destination: Aleixo Olímpico-e247145.mp3
[download] 100% of 3.58MiB in 00:05
[download] Downloading video 3 of 32
...
```

What else is necessary to merge?

File: youtube_dl/extractor/rtp.py [hunk 1]
```diff
@@ -87,3 +62,32 @@ def _real_extract(self, url):
             'description': description,
             'thumbnail': thumbnail,
         }
+
+
+class RTPPlaylistIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?rtp\.pt/play/p(?P<program_id>[0-9]+)(?!.+)'
```
Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: No trailing markers - override `suitable`.

File: youtube_dl/extractor/rtp.py [hunk 2]
```diff
@@ -87,3 +62,32 @@ def _real_extract(self, url):
...
+
+    def _get_program_id(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        program_id = mobj.group('program_id')
+        return program_id
+
+    def _extract_entries(self, url, program_id, page):
+        entry_url = "https://www.rtp.pt/play/bg_l_ep/?&listProgram=%s&listcategory=&listchannel=&type=radio&page=%s" % (
```
Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: Query must go to `query`.

File: youtube_dl/extractor/rtp.py [hunk 3]
```diff
@@ -29,55 +32,27 @@ class RTPIE(InfoExtractor):
...
         config = self._parse_json(player_config, video_id)
-
         path, ext = config.get('file').rsplit('.', 1)
         formats = [{
             'format_id': 'rtmp',
             'ext': ext,
-            'vcodec': config.get('type') == 'audio' and 'none' or None,
             'preference': -2,
-            'url': 'rtmp://{streamer:s}/{application:s}'.format(**config),
+            'url': '{file}'.format(**config),
```
Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: format is pointless.

File: youtube_dl/extractor/rtp.py [hunk 4]
```diff
@@ -29,55 +32,27 @@ class RTPIE(InfoExtractor):
...
-            },
-        }
-        r = replacements[config['type']]
-        if re.match(r['pattern'], config['file']) is not None:
-            formats.append({
-                'format_id': r['format_id'],
-                'url': re.sub(r['pattern'], r['repl'], config['file']),
-                'vcodec': r['vcodec'],
-            })
-
```
Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: Videos formats are not handled.

File: youtube_dl/extractor/rtp.py [hunk 5]
```diff
@@ -87,3 +62,32 @@ def _real_extract(self, url):
...
+                for episode in re.findall(r'e\d+', webpage)]
+
+    def _real_extract(self, url):
+        page = 1
+        program_id = self._get_program_id(url)
+
+        entry = self._extract_entries(url, program_id, page)
+        new_entry = self._extract_entries(url, program_id, page + 1)
+        while new_entry != []:
+            new_entry = self._extract_entries(url, program_id, page + 1)
```
Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: 87-90 code duplication.

Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: Also use generators or PagedList instead of list of entries.

File: youtube_dl/extractor/rtp.py [hunk 6]
```diff
@@ -87,3 +62,32 @@ def _real_extract(self, url):
...
+        program_id = mobj.group('program_id')
+        return program_id
+
+    def _extract_entries(self, url, program_id, page):
+        entry_url = "https://www.rtp.pt/play/bg_l_ep/?&listProgram=%s&listcategory=&listchannel=&type=radio&page=%s" % (
+            program_id, page
+        )
+        webpage = self._download_webpage(entry_url, program_id)
+        return [self.url_result('https://www.rtp.pt/play/p%s/%s/' % (program_id, episode), 'RTP')
+                for episode in re.findall(r'e\d+', webpage)]
```
Comment (dstftw) [inline youtube_dl/extractor/rtp.py]: Too broad regex.

**Comment by nimeir** [comment]
Thank you for the feedback! I'm working on it right now.

**Pull request closed without merging**